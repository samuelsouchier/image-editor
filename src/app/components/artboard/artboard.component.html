<section class="artboard" #artboard>
  <button class="artboard-backdrop" (click)="clearSelectedWidget()" (keyup.enter)="clearSelectedWidget()"> </button>
  <div
    *ngFor="let image of images(); trackBy:widgetById"
    class="artboard-widget-wrapper"
    cdkDrag
    cdkDragBoundary=".artboard"
    [ngStyle]="{
        'top': image.position.y + 'px',
        'left': image.position.x + 'px',
      }"
  >
    <button
      class="artboard-widget"
      [ngClass]="{selected: selectedWidget()?.id === image.id}"
      [ngStyle]="{
        'transform': 'rotate(' + image.rotation +'deg)',
        'height': image.height + 'px',
      }"
      (click)="selectWidget(image)"
      (keyup.enter)="selectWidget(image)"
    >
      <img
        [src]="image.src"
        [alt]="image.name"
        [ngStyle]="{
        'filter':
          'brightness(' + image.filters.brightness + ') ' +
          'blur(' + image.filters.blur + 'px) ' +
          'sepia(' + image.filters.sepia + ') ' +
          'grayscale(' + image.filters.grayscale + ') ' +
          'contrast(' + image.filters.contrast + ')',
      }"
      >
      <button class="artboard-widget-action artboard-widget-move" cdkDragHandle>
        <app-icon iconName="move" />
      </button>
      <button class="artboard-widget-action artboard-widget-resize" cdkDragHandle>
        <app-icon iconName="maximize" />
      </button>
    </button>
  </div>
</section>
