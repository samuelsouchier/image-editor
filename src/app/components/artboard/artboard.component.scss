@import 'src/styles/variables';
@import 'src/styles/colors';

:host {
  flex: 1;
}

.artboard {
  border-radius: $border-radius-l;
  border: 1px solid rgba($color-primary, .4);
  position: relative;
  height: 100%;
  background: rgb(246, 246, 246);
  background: linear-gradient(120deg, rgb(227, 227, 232) 0%, rgb(245, 245, 250) 19%, rgb(229, 229, 234) 45%, rgb(241, 241, 246) 78%, rgb(231, 231, 236) 100%);
  box-shadow: inset 0 0 12px 2px rgba($color-primary, .2);
  overflow: hidden;
  &-backdrop {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background: transparent;
    border: none;
    display: block;
  }
  &-widget {
    position: absolute;
    display: flex;
    overflow: hidden;
    border: 2px solid transparent;
    transition: box-shadow .3s ease-in, border .2s ease-in;
    cursor: pointer;
    padding: 0;
    background: transparent;
    height: fit-content;
    width: fit-content;
    resize: vertical;
    &-wrapper {
      height: fit-content;
      width: fit-content;
      position: absolute;
    }
    img {
      height: 100%;
      width: auto;
    }
    &-action {
      position: absolute;
      padding: $spacing-xs;
      background: transparent;
      transition: .2s ease-in;
      cursor: pointer;
      color: transparent;
      border: none;
      display: flex;
      justify-content: center;
      align-items: center;
      transform: scale(0.5);
    }
    &-move {
      transform-origin: top left;
      left: 0;
      top: 0;
      border-radius: 0 0 $border-radius 0;
    }
    &-resize {
      transform-origin: bottom right;
      bottom: 0;
      right: 0;
      border-radius: $border-radius 0 0 0;
      pointer-events: none;
      app-icon {
        transform: rotate(90deg);
      }
    }
    &:hover,
    &.selected {
      box-shadow: 0 2px 6px 2px rgba($color-primary, .2);
      border: 2px solid $color-primary;
      .artboard-widget-action {
        color: $color-white;
        background: $color-primary;
        &:active {
          cursor: grabbing;
          background: rgba($color-primary, 0.7);
        }
      }
    }

  }
}
