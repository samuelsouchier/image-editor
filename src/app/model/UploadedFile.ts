export interface UploadedFile {
  id: string,
  file: File,
}
