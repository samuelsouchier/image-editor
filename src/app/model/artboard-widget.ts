export interface ArtboardWidgetPosition {
  x: number,
  y: number,
}

export interface ArtboardWidgetImage {
  id: string;
  src: string;
  name: string;
  position: ArtboardWidgetPosition;
  height: number;
}
