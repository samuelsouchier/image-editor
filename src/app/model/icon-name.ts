import { blackWhiteIcon } from '../../assets/icons/black-white';
import { blurIcon } from '../../assets/icons/blur';
import { brightnessIcon } from '../../assets/icons/brightness';
import { clearIcon } from '../../assets/icons/clear';
import { contrastIcon } from '../../assets/icons/contrast';
import { cropIcon } from '../../assets/icons/crop';
import { imageIcon } from '../../assets/icons/image';
import { maximizeIcon } from '../../assets/icons/maximize';
import { moveIcon } from '../../assets/icons/move';
import { rotateBackIcon } from '../../assets/icons/rotate-back';
import { rotateForwardIcon } from '../../assets/icons/rotate-forward';
import { sepiaIcon } from '../../assets/icons/sepia';
import { trashIcon } from '../../assets/icons/trash';
import { typeIcon } from '../../assets/icons/type';
import { zoomInIcon } from '../../assets/icons/zoom-in';
import { zoomOutIcon } from '../../assets/icons/zoom-out';

export type IconName =
  'image'
  | 'trash'
  | 'crop'
  | 'type'
  | 'maximize'
  | 'move'
  | 'clear'
  | 'rotateBack'
  | 'rotateForward'
  | 'brightness'
  | 'blur'
  | 'contrast'
  | 'blackWhite'
  | 'sepia'
  | 'zoomIn'
  | 'zoomOut'
  ;
export const iconByNameMapper: Record<IconName, string> = {
  'image': imageIcon,
  'trash': trashIcon,
  'type': typeIcon,
  'crop': cropIcon,
  'maximize': maximizeIcon,
  'move': moveIcon,
  'clear': clearIcon,
  'rotateBack': rotateBackIcon,
  'rotateForward': rotateForwardIcon,
  'brightness': brightnessIcon,
  'blur': blurIcon,
  'contrast': contrastIcon,
  'blackWhite': blackWhiteIcon,
  'sepia': sepiaIcon,
  'zoomIn': zoomInIcon,
  'zoomOut': zoomOutIcon,
};
