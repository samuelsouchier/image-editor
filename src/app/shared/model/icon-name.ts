import { cropIcon } from '../../../assets/icons/crop';
import { imageIcon } from '../../../assets/icons/image';
import { trashIcon } from '../../../assets/icons/trash';
import { typeIcon } from '../../../assets/icons/type';

export type IconName = 'image' | 'trash' | 'crop' | 'type';
export const iconByNameMapper: Record<IconName, string> = {
  'image': imageIcon,
  'trash': trashIcon,
  'type': typeIcon,
  'crop': cropIcon,
};
