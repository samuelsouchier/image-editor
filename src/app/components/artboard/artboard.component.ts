import { CdkDrag, CdkDragHandle } from '@angular/cdk/drag-drop';
import { CommonModule } from '@angular/common';
import { Component, ElementRef, Input, OnChanges, SimpleChanges, ViewChild } from '@angular/core';
import { BASE_IMAGE_HEIGHT } from '../../constants';
import { ArtboardWidgetImage } from '../../model/artboard-widget';
import { WidgetStore } from '../../stores/widget.store';
import { IconComponent } from '../icon/icon.component';
import { ToolbarComponent } from '../toolbar/toolbar.component';

@Component({
  selector: 'app-artboard',
  standalone: true,
  imports: [CommonModule, CdkDrag, CdkDragHandle, IconComponent, ToolbarComponent],
  templateUrl: './artboard.component.html',
  styleUrls: ['./artboard.component.scss']
})
export class ArtboardComponent implements OnChanges {
  @Input() selectedFile: File | null;

  images = this.widgetStore.images;
  selectedWidget = this.widgetStore.selectedWidget;

  @ViewChild('artboard') artboard: ElementRef;

  constructor(private widgetStore: WidgetStore) {}

  ngOnChanges(changes: SimpleChanges) {
    if (changes['selectedFile']) {
      this.selectFile(this.selectedFile);
    }
  }

  widgetById(_: number, widget: ArtboardWidgetImage) {
    return widget.id;
  }

  selectWidget(widget: ArtboardWidgetImage) {
    this.widgetStore.selectWidget(widget);
  }

  selectFile(file: File | null) {
    if (file) {
      const reader = new FileReader();
      if (file) {
        reader.addEventListener('load', (event: ProgressEvent<FileReader>) => {
          if (typeof event.target?.result === "string") {
            const hostPosition = this.artboard.nativeElement.getBoundingClientRect();
            const image: ArtboardWidgetImage = {
              id: crypto.randomUUID(),
              src: event.target?.result,
              name: file.name,
              position: {
                x: hostPosition.width / 3,
                y: hostPosition.height / 3,
              },
              height: BASE_IMAGE_HEIGHT,
              rotation: 0,
              filters: {
                brightness: 1,
                blur: 0,
                contrast: 1,
              }
            };
            this.images.update(images => [...images, image]);
          }
        });

        reader.readAsDataURL(file);
      }
    }
  }
}
