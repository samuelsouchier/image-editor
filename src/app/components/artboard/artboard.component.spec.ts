import { signal, WritableSignal } from '@angular/core';
import { createComponentFactory, mockProvider, Spectator } from '@ngneat/spectator/jest';
import { artboardWidgetImageFixture } from '../../../fixtures/artboard-widget-image';
import { UploadedFile } from '../../model/UploadedFile';
import { ArtboardWidgetImage } from '../../model/artboard-widget';
import { WidgetStore } from '../../stores/widget.store';

import { ArtboardComponent } from './artboard.component';

describe(ArtboardComponent.name, () => {
  const images = signal<ArtboardWidgetImage[]>([artboardWidgetImageFixture]);
  const selectedWidget = signal<ArtboardWidgetImage | null>(artboardWidgetImageFixture);
  const selectedFile: WritableSignal<UploadedFile | null> = signal(null);
  let spectator: Spectator<ArtboardComponent>;
  const createComponent = createComponentFactory({
    component: ArtboardComponent,
    providers: [
      mockProvider(WidgetStore, {
        images,
        selectedFile,
        selectedWidget,
        updateWidget: (newWidget: ArtboardWidgetImage) => selectedWidget.set(newWidget),
        clearSelectedWidget: () => selectedWidget.set(null),
      }),
    ],
    shallow: true,
  });

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should match snapshot', () => {

    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should select a widget', () => {
    const mockWidget = artboardWidgetImageFixture;
    spectator.component.selectWidget(mockWidget);

    expect(selectedWidget()).toEqual(mockWidget);
  });

  it('should clear selected widget', () => {
    const mockWidget = artboardWidgetImageFixture;
    const expected = null;

    expect(selectedWidget()).toEqual(mockWidget);
    spectator.component.clearSelectedWidget();

    expect(selectedWidget()).toEqual(expected);
  });
});
