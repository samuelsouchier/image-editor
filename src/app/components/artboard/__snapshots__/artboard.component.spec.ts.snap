// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtboardComponent should match snapshot 1`] = `
<app-artboard
  artboard={[Function _ElementRef]}
  images={[Function Function]}
  selectedFile={[Function Function]}
  selectedWidget={[Function Function]}
  widgetStore={[Function Object]}
>
  <section
    class="artboard"
  >
    <button
      class="artboard-backdrop"
    />
    <div
      cdkdrag=""
      cdkdragboundary=".artboard"
      class="cdk-drag artboard-widget-wrapper"
      style="top: 150px; left: 150px;"
    >
      <button
        class="artboard-widget selected"
        style="transform: rotate(0deg) scale(1); height: 300px;"
      >
        <img
          alt="image-name"
          src="image-src"
          style="filter: brightness(1) blur(0px) sepia(0) grayscale(0) contrast(1);"
        />
        <button
          class="artboard-widget-action artboard-widget-move"
        >
          <app-icon
            iconname="move"
          >
            <div
              class="icon"
            >
              <svg
                class="feather feather-move"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="5 9 2 12 5 15"
                />
                <polyline
                  points="9 5 12 2 15 5"
                />
                <polyline
                  points="15 19 12 22 9 19"
                />
                <polyline
                  points="19 9 22 12 19 15"
                />
                <line
                  x1="2"
                  x2="22"
                  y1="12"
                  y2="12"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="2"
                  y2="22"
                />
              </svg>
            </div>
          </app-icon>
        </button>
      </button>
    </div>
  </section>
</app-artboard>
`;
